### PR TITLE
Cross-build test/images/serve_hostname

### LINKED ISSUE #31238: [test/images] cross-build test/images/serve_hostname
cross-build `test/images/serve_hostname`, https://github.com/kubernetes/kubernetes/tree/master/test/images/serve_hostname

Supported arch are:
* amd64(default)
* arm
* arm64
* ppc64le

PR is #31241.

This issue and PR(#31241) is an effort to achieve multiarch Kubernetes(#26863)

@luxas @spxtr

### PR DESCRIPTION
**Pull request by m1093782566**
**What this PR does / why we need it**:

**cross build** `test/images/serve_hostname`

https://github.com/kubernetes/kubernetes/tree/master/test/images/serve_hostname

This PR is an effort to achieve multiarch Kubernetes(#26863)

**Which issue this PR fixes** : 

fixes #31238 

@luxas @spxtr 

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31241)

### CODE REVIEW

File: test/images/serve_hostname/Makefile [hunk 1]
```diff
@@ -12,17 +12,45 @@
...
+#
+# Usage:
+#       [TAG=v1.4] [PREFIX=gcr.io/google_containers] [ARCH=amd64] make container
+
 all: serve_hostname
 
-TAG = v1.4
-PREFIX = gcr.io/google_containers
-TEST_PREFIX = b.gcr.io/k8s_authenticated_test
+TAG?=v1.4
```
Comment (luxas) [inline test/images/serve_hostname/Makefile]: tag should be updated to v1.5

**Comment by luxas** [comment]
ref: https://github.com/kubernetes/kubernetes/blob/master/test/images/serve_hostname/Dockerfile#L15 
the FROM statement should be replaced dynamically like etcd, hyperkube or a lot other images do

File: test/images/serve_hostname/Makefile [hunk 2]
```diff
@@ -12,17 +12,45 @@
...
+ARCH?=amd64
+PREFIX?=gcr.io/google_containers
+TEST_PREFIX=b.gcr.io/k8s_authenticated_test
+GOARM=6
+TEMP_DIR:=$(shell mktemp -d)
+KUBE_CROSS_TAG=$(shell cat ../../../build/build-image/cross/VERSION)
+
+ifeq ($(ARCH),amd64)
```
Comment (luxas) [inline test/images/serve_hostname/Makefile]: CC isn't needed at all here, use these for BASEIMAGE instead

File: test/images/serve_hostname/Makefile [hunk 3]
```diff
@@ -12,17 +12,45 @@
...
+        CC=powerpc64le-linux-gnu-gcc
+endif
 
 serve_hostname: serve_hostname.go
-	CGO_ENABLED=0 go build -a -installsuffix cgo --ldflags '-w' ./serve_hostname.go
+	# Copy the content in this dir to the temp dir
+	cp ./* $(TEMP_DIR)
+
+	docker run -it -v $(TEMP_DIR):/serve_hostname gcr.io/google_containers/kube-cross:$(KUBE_CROSS_TAG) /bin/bash -c \
+	    "cd /serve_hostname \
```
Comment (luxas) [inline test/images/serve_hostname/Makefile]: you can use `-w /serve_hostname` instead

File: test/images/serve_hostname/Makefile [hunk 4]
```diff
@@ -12,17 +12,45 @@
...
+	    && CGO_ENABLED=0 GOARM=$(GOARM) GOARCH=$(ARCH) CC=$(CC) CGO_ENABLED=0 go build -a -installsuffix cgo --ldflags '-w' /serve_hostname/serve_hostname.go"
+
 
 container: serve_hostname
-	docker build -t $(PREFIX)/serve_hostname:$(TAG) .
+	docker build -t $(PREFIX)/serve_hostname:$(TAG) $(TEMP_DIR)
```
Comment (luxas) [inline test/images/serve_hostname/Makefile]: This does still only build for one arch.
It should be `$(PREFIX)/serve_hostname-$(ARCH):$(TAG)`

**Comment by luxas** [comment]
cc @thockin @ixdy @AlexMioMio

**Comment by thockin** [comment]
Can we just steal the Makefile/Dockerfile/gitignore from https://github.com/kubernetes/kubernetes/tree/master/build/pause and convert it to build go instead of C?

**Comment by luxas** [comment]
@m1093782566 You should just steal this file (https://github.com/kubernetes/contrib/blob/master/404-server/Makefile), and only replace some values.

I'd like to get the Makefiles somehow standardized, so reusing an existing Makefile is way better than trying to write a new.

**Comment by thockin** [comment]
the pause one ensures that repeated `make` runs don't duplicate work

File: test/images/serve_hostname/Makefile (hunk 1 above)
Comment (m1093782566) [inline test/images/serve_hostname/Makefile]: updated to v1.5

File: test/images/serve_hostname/Makefile (hunk 2 above)
Comment (m1093782566) [inline test/images/serve_hostname/Makefile]: ok, removed CC and add BASEIMAGE.

File: test/images/serve_hostname/Makefile (hunk 3 above)
Comment (m1093782566) [inline test/images/serve_hostname/Makefile]: replaced with:

'-w' -o serve_hostname ./serve_hostname.go

File: test/images/serve_hostname/Makefile (hunk 4 above)
Comment (m1093782566) [inline test/images/serve_hostname/Makefile]: oh, sure! modified.

**Comment by m1093782566** [comment]
@luxas @thockin Thanks for your advice! PR is updated.

Thanks!

**Comment by m1093782566** [comment]
@luxas PR is updated. Would you please review it again?

Thanks!

**Comment by ixdy** [comment]
@m1093782566 I'm starting to suspect we may want to figure out some reusable make rules so we don't keep duplicating this logic in every image Makefile.

File: test/images/serve_hostname/README.md [hunk 5]
```diff
@@ -1,7 +1,31 @@
...
+
+The `serve_hostname` Makefile supports multiple architectures, which means it may cross-compile and build an docker image easily.
+Arch-specific busybox images serve as base images.
+
+If you are releasing a new version, please bump the `TAG` value in the `Makefile` before building the images.
+
+## How to release:
+
+```
+# Build for linux/amd64 (default)
```
Comment (ixdy) [inline test/images/serve_hostname/README.md]: maybe have `all-container` and `all-push` rules like the `pause` image uses?

**Comment by thockin** [comment]
You didn't take some of the interesting parts of the Makefile.

File: test/images/serve_hostname/README.md (hunk 5 above)
Comment (m1093782566) [inline test/images/serve_hostname/README.md]: ok, thanks for your advice.

**Comment by m1093782566** [comment]
@thockin I will try to take the interesting parts of the `pause` Makefile.

Thanks!

**Comment by m1093782566** [comment]
@ixdy 

>I'm starting to suspect we may want to figure out some reusable make rules so we don't keep duplicating this logic in every image Makefile.

I agree with you and hope can do some help to achieve this.

Thanks!

**Comment by luxas** [comment]
ping @m1093782566 Can you integrate with the pause image and I'll review?

**Comment by m1093782566** [comment]
@luxas 

Sure. I have updated this PR and take some "interesting" parts from `pause` Makefile as @thockin and @ixdy suggested. PTAL.

Thanks!

**Comment by m1093782566** [comment]
ping @luxas 

Comments addressed. PTAL.

**Comment by luxas** [comment]
ok to test

File: test/images/serve_hostname/Makefile [hunk 6]
```diff
@@ -12,26 +12,90 @@
...
+all-push: $(addprefix sub-push-,$(ALL_ARCH))
+
+build: bin/$(BIN)-$(ARCH)
+
+bin/$(BIN)-$(ARCH): $(SRCS)
+	# Copy the content in this dir to the temp dir
+	cp ./* $(TEMP_DIR)
+
+	docker run -it -v $(TEMP_DIR):/build \
+	$(KUBE_CROSS_IMAGE):$(KUBE_CROSS_VERSION) \
```
Comment (luxas) [inline test/images/serve_hostname/Makefile]: golang:$(GOLANG_VERSION) is just enough here, you can set GOLANG_VERSION to 1.6.3

File: test/images/serve_hostname/Makefile [hunk 7]
```diff
@@ -12,26 +12,90 @@
...
+            CGO_ENABLED=0 GOARM=$(GOARM) GOARCH=$(ARCH) go build -a -installsuffix cgo --ldflags '-w' -o $(BIN) ./$(SRCS)"
+
+container: .container-$(ARCH)
+.container-$(ARCH): bin/$(BIN)-$(ARCH)
+	# Set the base image
+	cd $(TEMP_DIR) && sed -i.bak 's|BASEIMAGE|$(BASEIMAGE)|g' Dockerfile
+
+	docker build -t $(IMAGE):$(TAG) $(TEMP_DIR)
+	if [ -n "$(TEST_REGISTRY)" ]; then \
+		docker tag -f $(IMAGE):$(TAG) $(TEST_IMAGE):$(TAG) ;\
```
Comment (luxas) [inline test/images/serve_hostname/Makefile]: remove `-f`

File: test/images/serve_hostname/README.md [hunk 8]
```diff
@@ -1,7 +1,31 @@
...
+$ make push
```
Comment (luxas) [inline test/images/serve_hostname/README.md]: Document here that one should just run `all-container` or if it's a release; run `all-push`

**Comment by luxas** [comment]
Three nits; otherwise LGTM

File: test/images/serve_hostname/Makefile (hunk 6 above)
Comment (m1093782566) [inline test/images/serve_hostname/Makefile]: Okay, will do.

File: test/images/serve_hostname/README.md (hunk 8 above)
Comment (m1093782566) [inline test/images/serve_hostname/README.md]: Reasonable, will do.

File: test/images/serve_hostname/Makefile (hunk 7 above)
Comment (m1093782566) [inline test/images/serve_hostname/Makefile]: Okay, will do.

**Comment by m1093782566** [comment]
ping @luxas comments addressed. PTAL

**Comment by luxas** [comment]
Thanks, LGTM!

**Pull request merged**